apps/heft/src/cli/actions/HeftActionBase.ts

```diff
@@ -17,20 +17,24 @@ import { PluginManager } from '../../pluginFramework/PluginManager';
 /**
  * @public
  */
-export interface IActionDataBase<THooks extends ActionHooksBase> {
+export interface IActionContext<
+  THooks extends ActionHooksBase<TActionProperties>,
+  TActionProperties extends object
+> {
   hooks: THooks;
+  properties: TActionProperties;
 }
 
 /**
  * @public
  */
-export abstract class ActionHooksBase {
+export abstract class ActionHooksBase<TActionProperties extends object> {
   /**
    * This hook allows the action's execution to be completely overridden. Only the last-registered plugin
    * with an override hook provided applies.
    */
-  public readonly override: AsyncSeriesBailHook<IActionDataBase<ActionHooksBase>> = new AsyncSeriesBailHook([
-    'actionData'
+  public readonly override: AsyncSeriesBailHook<TActionProperties> = new AsyncSeriesBailHook([
+    'actionContext'
   ]);
```

Comment:
The `AsyncSeriesBailHook<TActionProperties>` construct seems to produce an event handler callback signature like this:

```ts
call(arg1?: TActionProperties | undefined, arg2?: any, arg3?: any, ...args: any[]): any;
```

That might be the abstraction that tappable's pipeline invokes, but is there some way we could replace it in our own API with something stricter and intelligently named?

Something like this:
```ts
type OverrideHookCall<TActionProperties extends object> = (actionProperties: TActionProperties) => void;
```

Also, if this is an extensible event handler, the industry standard is more like this:
```ts
type EventHandler<TEventArgs> = (thingTheEventCameFrom: SomeBaseClass, eventArgs: TEventArgs) => void;

call: EventHandler<IOverrideHookEventArgs>;
```

And then `IOverrideHookEventArgs` is a documented and extensible.

Reply:
There's probably different ways to approach it. But the main feedback is that this isn't a great IntelliSense experience:

![Capture](https://user-images.githubusercontent.com/4673363/85791854-6071e900-b6e7-11ea-957f-78d8ec19b75f.png)

Reply:
Let's mark this as `@beta` for now and tackle this later as the API matures.